### PR TITLE
Modify API and Generate Addon API clientset/lister/informers

### PR DESCRIPTION
**Pull request by ccfishk**
Signed-off-by: jhu02 <jiminh@jiminhs-MacBook-Pro.local>

Fixes: 
https://github.com/keikoproj/addon-manager/issues/97
https://github.intuit.com/kubernetes/arktika/issues/3541

1. Use code-generator generate API(s) clientset/informer/listers and put into pkg/client directory. 
2. Add code-generator into Makefile
3. move api/v1alpha1 to apis/addon/v1alpha1, and add "comment marker" for code generator

### CODE REVIEW

File: pkg/apis/addon/v1alpha1/addon_types.go [hunk 1]
```diff
@@ -292,6 +294,8 @@ type AddonStatus struct {
 // +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.lifecycle.installed"
 // +kubebuilder:printcolumn:name="REASON",type="string",JSONPath=".status.reason"
 // +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +genclient
```
Comment (kevdowney) [inline pkg/apis/addon/v1alpha1/addon_types.go:297]: Addon-manager and most other Keikoproj controllers were generated by **kubebuilder@v2** and according to the migration [docs](https://book.kubebuilder.io/migration/legacy/migration_guide_v1tov2.html#migrate-the-apis) 

> We don’t need the following markers any more (they’re not used anymore, and are relics from much older versions of KubeBuilder):
// +genclient
// +k8s:openapi-gen=true

Addon-manager should remain aligned to v2 of Kubebuilder until we upgrade to v3, but why are these comment markers?
 
I also have concern about removing `/api` directory as that is required by **kubebuilder** scaffolding

**Review by figo — CHANGES_REQUESTED** [review]
Thank you for working on this refactoring effort, Refactor is good, but would like to share a few guideline to help both reviewers and PR owners.
1. In the PR description, document what problem the PR trying to solve (break the large PR into smaller ones, and make it incremental as much as possible), "refactoring" is the approach, but it is not a problem statement, 
2. To make sure reviewer can review the PR thoroughly, add the dependencies and steps that generate the PR into a script (in this case, kubebuilder version etc), include the script in the PR, this will make sure others can also generate the PR using the script, this apply to PR when there is generated code.
3. (optional) walk through the PR with reviewers, and share the plan for the following PRs.

Thanks again.

File: pkg/apis/addon/v1alpha1/addon_types.go (hunk 1 above)
Comment (ccfishk) [reply pkg/apis/addon/v1alpha1/addon_types.go:297]: Those "comment marker" works for code-generator. And it does not impact kubebuilder. Believing as long as we do not remove kubebuilder comments, kubebuilder v2/v3 should work as expected. 
both code-generator and kubebuilder has options specifying API directory, do not see any concern here.

Comment (kevdowney) [reply pkg/apis/addon/v1alpha1/addon_types.go:297]: In that case not sure why we'd choose to remove `/api` if you can generate to any directory, the design appears to resemble [sample-controller](https://github.com/kubernetes/sample-controller/tree/master/pkg/apis/samplecontroller/v1alpha1) which pre-dates **kubebuilder**, likely the genesis of kubebuilder@v0 and we're in v2.

I don't see the reason to remove `/api` even if you want to just have `/pkg/apis`.

**Comment by ccfishk** [comment]
> Thank you for working on this refactoring effort, Refactor is good, but would like to share a few guideline to help both reviewers and PR owners.
> 
> 1. In the PR description, document what problem the PR trying to solve (break the large PR into smaller ones, and make it incremental as much as possible), "refactoring" is the approach, but it is not a problem statement,
> 2. To make sure reviewer can review the PR thoroughly, add the dependencies and steps that generate the PR into a script (in this case, kubebuilder version etc), include the script in the PR, this will make sure others can also generate the PR using the script, this apply to PR when there is generated code.
> 3. (optional) walk through the PR with reviewers, and share the plan for the following PRs.
> 
> Thanks again.
This comments justifying the strategy
https://github.com/keikoproj/addon-manager/pull/99#issuecomment-1046303380

**Comment by figo** [comment]
> > Thank you for working on this refactoring effort, Refactor is good, but would like to share a few guideline to help both reviewers and PR owners.
> > 
> > 1. In the PR description, document what problem the PR trying to solve (break the large PR into smaller ones, and make it incremental as much as possible), "refactoring" is the approach, but it is not a problem statement,
> > 2. To make sure reviewer can review the PR thoroughly, add the dependencies and steps that generate the PR into a script (in this case, kubebuilder version etc), include the script in the PR, this will make sure others can also generate the PR using the script, this apply to PR when there is generated code.
> > 3. (optional) walk through the PR with reviewers, and share the plan for the following PRs.
> > 
> > Thanks again.
> > This comments justifying the strategy
> > [#99 (comment)](https://github.com/keikoproj/addon-manager/pull/99#issuecomment-1046303380)

Yes, glad those 3 action items make sense to you, could you start to address them?

File: pkg/apis/addon/v1alpha1/addon_types.go (hunk 1 above)
Comment (ccfishk) [reply pkg/apis/addon/v1alpha1/addon_types.go:297]: Move api and client to a central place or not is not a blocker at all.  
Not quite get the comparison between kubebuilder and workqueue (sample, argo etc used), I do not think they have any conflicts.

Comment (kevdowney) [reply pkg/apis/addon/v1alpha1/addon_types.go:297]: Argo workflow-controller pre-dates **kubebuilder** and used **sample-controller** design, this is not the same as  **kubebuilder** which has an opinion about generated clientset... See https://github.com/kubernetes-sigs/kubebuilder/issues/403#issuecomment-518437978
> "Libraries Over Code Generation"

https://github.com/kubernetes-sigs/kubebuilder/blob/master/DESIGN.md#overarching

What we're saying is if we want to generate a clientset let's not change the opionated way of kubebuilder but instead do it separately; there's no requirement to remove `/api`.

Comment (kevdowney) [reply pkg/apis/addon/v1alpha1/addon_types.go:297]: You can have both `/api` and `/pkg/apis` as separate things, correct?

Comment (ccfishk) [reply pkg/apis/addon/v1alpha1/addon_types.go:297]: We use clientset/lister/informer accessing addon API and thus react (install workflow, update status) directly. with so, reduce cpu/mem, and process cycle.
We only need maintain one Addon API in the repo, we not allowed duplicate definition

**Comment by ccfishk** [comment]
> > > Thank you for working on this refactoring effort, Refactor is good, but would like to share a few guideline to help both reviewers and PR owners.
> > > 
> > > 1. In the PR description, document what problem the PR trying to solve (break the large PR into smaller ones, and make it incremental as much as possible), "refactoring" is the approach, but it is not a problem statement,
> > > 2. To make sure reviewer can review the PR thoroughly, add the dependencies and steps that generate the PR into a script (in this case, kubebuilder version etc), include the script in the PR, this will make sure others can also generate the PR using the script, this apply to PR when there is generated code.
> > > 3. (optional) walk through the PR with reviewers, and share the plan for the following PRs.
> > > 
> > > Thanks again.
> > > This comments justifying the strategy
> > > [#99 (comment)](https://github.com/keikoproj/addon-manager/pull/99#issuecomment-1046303380)
> 
> Yes, glad those 3 action items make sense to you, could you start to address them?

Documented the problem details in following ticket tracking all the solutions that are applied.
https://github.com/keikoproj/addon-manager/pull/99#issue-1134505126

File: pkg/apis/addon/v1alpha1/addon_types.go (hunk 1 above)
Comment (kevdowney) [reply pkg/apis/addon/v1alpha1/addon_types.go:297]: I am against moving things to `/pkg` if you change to use top-level `/apis` directory I will overlook the pathing.

For the rest of this PR I expect this code generator to be tied into the build, either via go generate or through the makefile.

See how someone else did similar.
https://www.fatalerrors.org/a/writing-crd-by-mixing-kubeuilder-and-code-generator.html

Comment (ccfishk) [reply pkg/apis/addon/v1alpha1/addon_types.go:297]: This is an awesome idea, will follow.

Comment (ccfishk) [reply pkg/apis/addon/v1alpha1/addon_types.go]: Fixed https://github.com/keikoproj/addon-manager/pull/111/commits/d05c872afa597828065cea15a2178ad981f65dd0

File: Makefile [hunk 2]
```diff
@@ -130,3 +130,10 @@ CONTROLLER_GEN=$(GOBIN)/controller-gen
 else
 CONTROLLER_GEN=$(shell which controller-gen)
 endif
+
+code-generator:
+bash $(GOPATH)/src/k8s.io/code-generator@v0.21.5/generate-groups.sh \
```
Comment (kevdowney) [inline Makefile:135]: Fix indentation, and does it really need bash here or just exection as `controller-gen` has it. Also how do I know I have this script locally?
```suggestion
    bash $(GOPATH)/src/k8s.io/code-generator@v0.21.5/generate-groups.sh \
```

Comment (ccfishk) [reply Makefile]: 2729a2449013ce01bb998ed635413d3b78585a2c fixed.
Yeah, the tool code-generator has to be exactly installed under go/src/k8s.io/ afaik.

File: Dockerfile [hunk 3]
```diff
@@ -12,7 +12,7 @@ ADD go.sum .
 RUN go mod download
 
 COPY pkg/ pkg/
-COPY api/ api/
+COPY apis/ apis
```
Comment (kevdowney) [inline Dockerfile:15]:
```suggestion
COPY apis/ apis/
```

File: Makefile [hunk 4]
```diff
@@ -26,8 +26,24 @@ GO111MODULE=on
 all: test manager addonctl
 
 # Run tests
-test: generate fmt vet manifests
```
Comment (kevdowney) [inline Makefile:29]: Does this change the report, I want to run all tests as one report.

Comment (ccfishk) [reply Makefile:29]: No, it does not. The report is the same.

File: Makefile [hunk 5]
```diff
@@ -130,3 +146,10 @@ CONTROLLER_GEN=$(GOBIN)/controller-gen
 else
 CONTROLLER_GEN=$(shell which controller-gen)
 endif
+
+code-generator:
+  bash $(GOPATH)/src/k8s.io/code-generator@v0.21.5/generate-groups.sh \
```
Comment (kevdowney) [inline Makefile:151]: This needs to be a tab indent. Code generator also needs to be added to the `make generate` cadence.

File: e2e/addon_types_test.go [hunk 6]
```diff
@@ -12,7 +12,7 @@
  * limitations under the License.
  */
 
-package v1alpha1
+package e2e
```
Comment (kevdowney) [inline e2e/addon_types_test.go:15]: why is this moved?

Comment (ccfishk) [reply e2e/addon_types_test.go:15]: API tests need refer API definition such as AddToScheme (BuilderScheme has been upgraded) to launch client, which is different from previous scheme. 
Previous API location introduced in loop.

File: Dockerfile (hunk 3 above)
Comment (ccfishk) [reply Dockerfile]: 28000e03e05610fb8b3149b1e5ef52fe239256a4

File: e2e/suite_test.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
+// These tests use Ginkgo (BDD-style Go testing framework). Refer to
+// http://onsi.github.io/ginkgo/ to learn more about Ginkgo.
+
+package e2e
```
Comment (kevdowney) [inline e2e/suite_test.go:4]: Why do we need another copy of controller `suite_test.go`?

Comment (ccfishk) [reply e2e/suite_test.go:4]: Though the setup is similar, but we do not want mix them. 
Also, I will add more e2e tests in the coming PR(s)

File: Makefile (hunk 4 above)
Comment (kevdowney) [reply Makefile:29]: Please post a sample of `make test` output

File: Makefile [hunk 8]
```diff
@@ -96,7 +96,7 @@ vet:
 
 # Generate code
 generate: controller-gen
```
Comment (kevdowney) [inline Makefile:114]: Add `code-generator` here as well.

File: Makefile (hunk 2 above)
Comment (kevdowney) [reply Makefile]: Fix tab indent and add download logic for this.

Comment (ccfishk) [reply Makefile]: Yes, it has been tab and add download task within Makefile. 
0ce90df6cde7f399781b319c16cfe1ce40cddfb6

File: Makefile (hunk 5 above)
Comment (ccfishk) [reply Makefile]: Added code-generator download task
https://github.com/keikoproj/addon-manager/pull/111/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R151

File: e2e/addon_types_test.go (hunk 6 above)
Comment (ccfishk) [reply e2e/addon_types_test.go]: Per discussion, move them back to apis/addon directory

0ce90df6cde7f399781b319c16cfe1ce40cddfb6

File: Makefile (hunk 8 above)
Comment (ccfishk) [reply Makefile:115]: added
https://github.com/keikoproj/addon-manager/pull/111/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R116

File: Makefile [hunk 9]
```diff
@@ -26,8 +26,25 @@ GO111MODULE=on
 all: test manager addonctl
 
 # Run tests
-test: generate fmt vet manifests
-	go test ./api/... ./controllers/... ./pkg/... ./cmd/... -coverprofile cover.out
+.PHONY: test
+test:
```
Comment (kevdowney) [inline Makefile:30]: Needs the other targets also.
```suggestion
test: generate fmt vet manifests
```

File: Makefile [hunk 10]
```diff
@@ -95,8 +112,13 @@ vet:
 	go vet ./...
 
 # Generate code
-generate: controller-gen
-	$(CONTROLLER_GEN) object:headerFile=./hack/boilerplate.go.txt paths=./api/...
+generate: controller-gen code-generator
+	$(CONTROLLER_GEN) object:headerFile=./hack/boilerplate.go.txt paths=./apis/addon/...
+	bash $(GOPATH)/src/k8s.io/code-generator@v0.21.5/generate-groups.sh \
```
Comment (kevdowney) [inline Makefile:117]: Let's capture this code-generator call as env. var, `$(CODE_GEN)`

File: Makefile [hunk 11]
```diff
@@ -130,3 +152,15 @@ CONTROLLER_GEN=$(GOBIN)/controller-gen
...
+
+code-generator:
+ifeq (, $(shell which code-generator))
+	@{ \
+	set -e ;\
+	CODE_GENERATOR_DIR=$$(mkdir -p $(GOPATH)/src/k8s.io/) ;\
+	cd $$CODE_GENERATOR_DIR ;\
+	go get github.com/kubernetes/code-generator@v0.21.5 ;\
+	}
+endif
```
Comment (kevdowney) [inline Makefile:164]:
```suggestion
else
CODE_GEN="bash $(GOPATH)/src/k8s.io/code-generator@v0.21.5/generate-groups.sh"
endif
```

**Comment by ccfishk** [comment]
per conversation, this PR is moved to https://github.com/keikoproj/addon-manager/pull/116.

**Comment by ccfishk** [comment]
close as it is.
